dashboard/api.py

```diff
@@ -532,10 +532,22 @@ def get_edx_exam_coupon_url(user, course):
         user=user,
         course=course,
         status=ExamAuthorization.STATUS_SUCCESS,
-        exam_run__in=schedulable_exam_runs,
-        exam_coupon_url__isnull=False
+        exam_run__in=schedulable_exam_runs
     ).first()
-    return exam_auth.exam_coupon_url if exam_auth else ""
+    if exam_auth is None:
+        return ""
+
+    exam_coupon = ExamRunCoupon.objects.filter(
+        expiration_date__gte=now.date(),
+        is_taken=False,
+        course=course
+    ).first()
+    if exam_coupon is None:
+        return ""
+
+    exam_auth.exam_coupon = exam_coupon
+    exam_auth.save()
+    return exam_coupon.use_coupon()
```

Comment:
I don't know if it's something to be considered but seems like a bug to me so mentioning it here.

Let's say we have 10 ExamRunCoupon with is_taken=False, when a user loads its dashboard we assign the first ExamRunCoupon to ExamAuthorization and toggle is_taken to true.
If the user loads the dashboard again, this code will pick the second ExamRunCoupon with is_taken=False and toggles that, in this way, all the coupons will be marked as is_taken=True while only 1 of them will be getting used.

In the second scenario, if a user loads his dashboard and the system assigns user's ExamAuthorization object with a ExamRunCoupon. Now let's assume there are no  ExamRunCoupon with is_taken=False are left, now if the user loads his dashboard the edx exam_url will be empty for him although he was assigned a coupon earlier.

Reply:
you are right, thanks for catching that.